Source/x11/XGServerWindow.m

```diff
@@ -4430,74 +4436,154 @@ - (void) freecursor: (void*) cid
   return depth;
 }
 
+/* This method assumes that we deal with one X11 screen - `defScreen`.
+   Basically it means that we have DISPLAY variable set to `:0.0`.
+   Where both digits have artbitrary values, but it defines once on
+   every application run.
+   AppKit and X11 use the same term "screen" with different meaning:
+   AppKit Screen - single monitor/display device.
+   X11 Screen - it's a plane where X Server can draw on.
+   XRandR - Xorg extension that helps to manipulate monitor layout providing
+   X11 Screen. 
+   We map XRandR monitors (outputs) to NSScreen. */
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSArray        *screens;
+  NSMutableArray *tmpScreens;
+  int            i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+    monitors = NULL;
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  RROutput            primary_output;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+  int                 mi;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  if (screen_res != NULL)
+    {
+      monitorsCount = screen_res->noutput;
+      tmpScreens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      primary_output = XRRGetOutputPrimary(dpy, [self xDisplayRootWindow]);
+      for (i = 0, mi = 0; i < screen_res->noutput; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+              
+              monitors[mi].screen_id = defScreen;
+              monitors[mi].depth = [self windowDepthForScreen: mi];
+              monitors[mi].resolution = [self resolutionForScreen: defScreen];
```

Comment:
You could compute this once before the loop. And it still looks wrong to me. Why would different monitors have the same resolution?

Reply:
Indeed it's wrong. But resolutionForScreen is not implemented yet correctly. It's always returns {72, 72}. I've already said - it's rather complicated separate task and involves -back and -gui changes. I'll change this after correct resolutionForScreen: implemenation.